### PR TITLE
sclang: fix hang on floating point arguments to Integer:forBy

### PR DESCRIPTION
**Pull request by lnihlen**
Regarding #2690.

Previously Integer:forBy would truncate floating point arguments, to integers,
which could potentially result in an infinite loop by the Integer:forBy
optimized special operation. This modification checks for any of the arguments
to be floating point, and if detecting that casts all of the arguments to
floating point. It also adds appropriate testing coverage.

### CODE REVIEW

File: lang/LangSource/PyrInterpreter3.cpp [hunk 1]
```diff
@@ -1652,31 +1652,45 @@ HOT void Interpret(VMGlobals *g)
...
+						slotCopy(&vars[4], &g->receiver);
 					}
-					slotCopy(&vars[4], &g->receiver);
 					dispatch_opcode;
 				}
 				case 8 : {
 					PyrSlot * vars = g->frame->vars;
-					if ((slotRawInt(&vars[2]) >= 0 && slotRawInt(&vars[4]) <= slotRawInt(&vars[1]))
-							|| (slotRawInt(&vars[2]) < 0 && slotRawInt(&vars[4]) >= slotRawInt(&vars[1]))) {
+					bool continue_for_by = false;
```
Comment (brianlheim) [inline lang/LangSource/PyrInterpreter3.cpp]: the codebase isn't super consistent but I'd say we generally prefer camelCase naming for local variables.

File: lang/LangSource/PyrInterpreter3.cpp [hunk 2]
```diff
@@ -1652,31 +1652,45 @@ HOT void Interpret(VMGlobals *g)
...
-					if (IsFloat(vars+2)) {
-						SetInt(&vars[2], (int32)(slotRawFloat(&vars[2])));
-					}
-					int tag = GetTag(&vars[1]);
-					if ((tag != tagInt)
-							|| NotInt(&vars[2])) {
-						error("Integer-forBy : endval or stepval not an Integer.\n");
+					// If any argument is floating point we cast all arguments
+					// to floating point, including the accumulator. This avoids
+					// potential infinite loops due to integer truncation.
```
Comment (brianlheim) [inline lang/LangSource/PyrInterpreter3.cpp]: :+1:

File: testsuite/sclang/TestInteger.sc [hunk 3]
```diff
@@ -0,0 +1,29 @@
+TestInteger : UnitTest {
+	test_forby_with_positive_integer_arguments {
+		var forby_arguments = List.new;
```
Comment (brianlheim) [inline testsuite/sclang/TestInteger.sc]: and in sclang code we definitely prefer camelCase

File: testsuite/sclang/TestInteger.sc [hunk 4]
```diff
@@ -0,0 +1,29 @@
+TestInteger : UnitTest {
+	test_forby_with_positive_integer_arguments {
```
Comment (brianlheim) [inline testsuite/sclang/TestInteger.sc]: It's not 100% standardized yet, but I've got a WIP document for the project test suite [here](https://gist.github.com/brianlheim/91222d487afa18582c287b0a722ae272). The naming convention I'm trying to adopt is `test_methodName_inputConditions[_expectedResult]`, so following that this test should be named something like `test_forBy_positiveIntegerArguments`. Also, you can omit the messages if you want, it's nice to have but IMO they're not super helpful beyond the names of the test functions.

File: lang/LangSource/PyrInterpreter3.cpp (hunk 1 above)
Comment (lnihlen) [reply lang/LangSource/PyrInterpreter3.cpp]: Done.

File: testsuite/sclang/TestInteger.sc (hunk 4 above)
Comment (lnihlen) [reply testsuite/sclang/TestInteger.sc]: Done, thanks for the pointer to the guide.

File: testsuite/sclang/TestInteger.sc (hunk 3 above)
Comment (lnihlen) [reply testsuite/sclang/TestInteger.sc]: Done.

**Comment by lnihlen** [comment]
Thanks for the feedback. I added error detection for the case of the zero stepval. There's still a possibility of infinite loop by underflow but with 64-bit doubles this seems less troublesome.

**Comment by brianlheim** [comment]
> I added error detection for the case of the zero stepval. There's still a possibility of infinite loop by underflow but with 64-bit doubles this seems less troublesome.

Thanks for that!

> Oh, on the testing - I looked but didn't see any way to assert in testing that a particular operation results in a language error. 

We have an open PR for this actually - #3808 . If you don't mind waiting a few days, it should be merged soon and then we rebase this PR and use it!

**Comment by lnihlen** [comment]
In the interest of expedience would it be reasonable for me to accept an open issue to add test coverage for the two failure cases (non-numeric arguments and zero stepval), with the agreement that I would put that up for review once #3808 is in?

**Comment by lnihlen** [comment]
I mean, and then merge this PR in the meantime?

**Comment by brianlheim** [comment]
> In the interest of expedience would it be reasonable for me to accept an open issue to add test coverage for the two failure cases (non-numeric arguments and zero stepval), with the agreement that I would put that up for review once #3808 is in?

Not sure. In general our workflow has been to make things like this blocking, which I think helps to cut down on bookkeeping (and because people sometimes forget or fail to follow through on promises). But you've been staying on top of this one and it already has decent tests so let's give it a try. :)

File: testsuite/classlibrary/TestInteger.sc [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+		forBy(-6, -4, 0.5, { | i, j | forByArguments.add([i, j]); });
+		this.assertEquals(forByArguments, List[[-6.0, 0], [-5.5, 1], [-5.0, 2], [-4.5, 3], [-4.0, 4]]);
+	}
+
+	test_forBy_negativeFloatArguments {
+		var forByArguments = List.new;
+		21.forBy(14.5, -3.125, { | i, j | forByArguments.add([i, j]); });
+		this.assertEquals(forByArguments, List[[21.0, 0], [17.875, 1], [14.75, 2]]);
+	}
+}
```
Comment (brianlheim) [inline testsuite/classlibrary/TestInteger.sc]: An unfortunate bug that hasn't been addressed - the IDE doesn't append a newline to the end of the final line of a document, so most files edited with it will appear to be missing a final newline in git. I looked into it a couple months ago but didn't really come up with anything, and ended up moving on to other issues. Something to keep in mind for the future

**Comment by lnihlen** [comment]
Thanks for the review. In general my inclination is to decouple PRs as much as possible, to avoid possible propagation of delay through the PR dependency chain. Very hand-wavy, and I can't come up with a more concrete reason, so I appreciate the accommodation.
